### PR TITLE
Add `jigsaw-python` as a submodule

### PR DESCRIPTION
**Pull request by xylar**
This is an updated version of #550 

Checklist
* [x] Developer's Guide has been updated
* [x] Documentation has been [built locally](https://mpas-dev.github.io/compass/latest/developers_guide/building_docs.html) and changes look as expected
* [x] Document (in a comment titled `Testing` in this PR) any testing that was used to verify the changes

### CODE REVIEW

**Comment by xylar** [comment]
## Testing

I was able to build a conda-environment with `jigsawpy` 1.0.0 in it using this branch on Chrysalis.  I was then able to run the QU240 mesh, init and performance tests cases with it.

**Comment by xylar** [comment]
I still need to document the `jigsaw` and `jigsaw-python` build process and the `--update_jigsaw` flag.

@cbegeman and @matthewhoffman, this is otherwise ready for you to test and review when you have time.

**Comment by xylar** [comment]
@cwsmith, please let me know if this works well for you as well.

**Comment by cbegeman** [comment]
@xylar Great! I'll try to test this out by the end of the week.

**Comment by cwsmith** [comment]
@xylar This worked for me.  Thank you!

**Review by cbegeman — APPROVED** [review]
@xylar Thanks for working on this! I successfully ran QUwISC240 mesh, init, performance on perlmutter with gnu, mpich. 

As I mentioned, the jigsaw-python directory was empty with the following commands in my base environment:
```
git worktree add --checkout ../add-jigsaw-as-submodule xylar/compass/add-jigsaw-as-submodule
git submodule update --init --recursive
```
but I had success after deleting the directory, git hard-reset, and git submodule update as you suggested.

**Comment by xylar** [comment]
@cbegeman, I ran a slightly different variant of your commands and it worked fine for me:
```
git worktree add ../add-jigsaw-as-submodule -b add-jigsaw-as-submodule --checkout xylar/compass/add-jigsaw-as-submodule
cd ../add-jigsaw-as-submodule/
git submodule update --init
```
I don't see any reason that my commands would work and yours would not.  Let's assume for now that it was a fluke and that things will work better in the future.  But we need to keep an eye on this to make sure it doesn't happen again.

**Review by matthewhoffman — CHANGES_REQUESTED** [review]
@xylar , sorry for the delay in looking this over.  I added a couple questions in comments.  I tried building a new compass env on Anvil but ran into an error.  It may be related to my home dir running out of space, so I will look into that closer tomorrow and also try on Perlmutter when it is back up.

File: docs/developers_guide/quick_start.rst [hunk 1]
```diff
@@ -214,14 +219,18 @@ Optional flags
 ``--python``
     Select a particular python version (the default is currently 3.8)
 
-``--env-name``
+``--env_name``
     Set the name of the environment (and the prefix for the activation script)
     to something other than the default (``dev_compass_<version>`` or
     ``dev_compass_<version>_<mpi>``).
 
-``--with-albany``
```
[the remark this answers is not in the document]
Comment (xylar) [reply docs/developers_guide/quick_start.rst:222]: This test was redundant with text above that describes the same flag.  Also, the flag is `--with_albany` with an underscore.

File: docs/developers_guide/quick_start.rst [hunk 2]
```diff
@@ -214,14 +219,18 @@ Optional flags
 ``--python``
     Select a particular python version (the default is currently 3.8)
 
-``--env-name``
+``--env_name``
```
[the remark this answers is not in the document]
Comment (xylar) [reply docs/developers_guide/quick_start.rst:222]: Yes, all flags to `compass` commands are with underscores, not dashes.

**Comment by xylar** [comment]
@matthewhoffman, I updated the text about Jigsaw as you suggested.  The other changes you flagged are intentional.

Let me know if you have more success with your testing later this week.

**Comment by matthewhoffman** [comment]
@xylar , I'm seeing the same problem on perlmutter as I saw on Anvil last night, and it sounds similar to Carolyn's experience.  Possibly I'm not following the correct workflow, though.  Here is what I did and the subsequent error:

```
git worktree add ../jigsaw_submodule
git fetch --all -p
git checkout xylar/add-jigsaw-as-submodule
git submodule update --init jigsaw-python
./conda/configure_compass_env.py --conda /global/cfs/cdirs/fanssie/users/hoffman2/mambaforge -c gnu --mpi mpich -m pm-cpu --with_albany

Logging to: conda/logs/prebootstrap.log

Doing initial setup

Updating conda environment for installing compass

Creating the compass conda environment


 Running:
   source /global/cfs/cdirs/fanssie/users/hoffman2/mambaforge/etc/profile.d/conda.sh
   conda activate compass_bootstrap
   /global/cfs/cdirs/fanssie/users/hoffman2/compass/jigsaw_submodule/conda/bootstrap.py --conda /global/cfs/cdirs/fanssie/users/hoffman2/mambaforge -c gnu --mpi mpich -m pm-cpu --with_albany

Logging to: conda/logs/bootstrap.log

Configuring environment(s) for the following compilers and MPI libraries:
  gnu, mpich

Updating dev_compass_1.2.0

Installing compass

Traceback (most recent call last):
  File "/global/cfs/cdirs/fanssie/users/hoffman2/compass/jigsaw_submodule/conda/bootstrap.py", line 1107, in <module>
    main()
  File "/global/cfs/cdirs/fanssie/users/hoffman2/compass/jigsaw_submodule/conda/bootstrap.py", line 1023, in main
    build_conda_env(
  File "/global/cfs/cdirs/fanssie/users/hoffman2/compass/jigsaw_submodule/conda/bootstrap.py", line 371, in build_conda_env
    check_call(commands, logger=logger)
  File "/global/cfs/cdirs/fanssie/users/hoffman2/compass/jigsaw_submodule/conda/shared.py", line 151, in check_call
    raise subprocess.CalledProcessError(process.returncode, commands)
subprocess.CalledProcessError: Command 'source /global/cfs/cdirs/fanssie/users/hoffman2/mambaforge/etc/profile.d/conda.sh &&conda activate dev_compass_1.2.0 && cd /global/cfs/cdirs/fanssie/users/hoffman2/compass/jigsaw_submodule && python -m pip install -e .' returned non-zero exit status 1.
Traceback (most recent call last):
  File "./conda/configure_compass_env.py", line 134, in <module>
    main()
  File "./conda/configure_compass_env.py", line 130, in main
    bootstrap(activate_install_env, source_path, local_conda_build)
  File "./conda/configure_compass_env.py", line 37, in bootstrap
    check_call(command)
  File "/global/cfs/cdirs/fanssie/users/hoffman2/compass/jigsaw_submodule/conda/shared.py", line 151, in check_call
    raise subprocess.CalledProcessError(process.returncode, commands)
subprocess.CalledProcessError: Command 'source /global/cfs/cdirs/fanssie/users/hoffman2/mambaforge/etc/profile.d/conda.sh && conda activate compass_bootstrap && /global/cfs/cdirs/fanssie/users/hoffman2/compass/jigsaw_submodule/conda/bootstrap.py --conda /global/cfs/cdirs/fanssie/users/hoffman2/mambaforge -c gnu --mpi mpich -m pm-cpu --with_albany' returned non-zero exit status 1.
```

when I run the individual commands in the failed command manually, I get the following error on the bootstrap command:
```
$ /global/cfs/cdirs/fanssie/users/hoffman2/compass/jigsaw_submodule/conda/bootstrap.py --conda /global/cfs/cdirs/fanssie/users/hoffman2/mambaforge -c gnu --mpi mpich -m pm-cpu --with_albany
Logging to: conda/logs/bootstrap.log

Configuring environment(s) for the following compilers and MPI libraries:
  gnu, mpich

Updating dev_compass_1.2.0

Installing compass

Traceback (most recent call last):
  File "/global/cfs/cdirs/fanssie/users/hoffman2/compass/jigsaw_submodule/conda/bootstrap.py", line 1107, in <module>
    main()
  File "/global/cfs/cdirs/fanssie/users/hoffman2/compass/jigsaw_submodule/conda/bootstrap.py", line 1023, in main
    build_conda_env(
  File "/global/cfs/cdirs/fanssie/users/hoffman2/compass/jigsaw_submodule/conda/bootstrap.py", line 371, in build_conda_env
    check_call(commands, logger=logger)
  File "/global/cfs/cdirs/fanssie/users/hoffman2/compass/jigsaw_submodule/conda/shared.py", line 151, in check_call
    raise subprocess.CalledProcessError(process.returncode, commands)
subprocess.CalledProcessError: Command 'source /global/cfs/cdirs/fanssie/users/hoffman2/mambaforge/etc/profile.d/conda.sh &&conda activate dev_compass_1.2.0 && cd /global/cfs/cdirs/fanssie/users/hoffman2/compass/jigsaw_submodule && python -m pip install -e .' returned non-zero exit status 1.
```

If I then run those commands manually, pip install fails like this:
```
$ python -m pip install -e .
Obtaining file:///global/cfs/cdirs/fanssie/users/hoffman2/compass/jigsaw_submodule
  Preparing metadata (setup.py) ... done
Requirement already satisfied: cartopy in /global/cfs/cdirs/fanssie/users/hoffman2/mambaforge/envs/dev_compass_1.2.0/lib/python3.10/site-packages (from compass==1.2.0a9) (0.22.0)
Requirement already satisfied: cmocean in /global/cfs/cdirs/fanssie/users/hoffman2/mambaforge/envs/dev_compass_1.2.0/lib/python3.10/site-packages (from compass==1.2.0a9) (3.0.3)
Requirement already satisfied: gsw in /global/cfs/cdirs/fanssie/users/hoffman2/mambaforge/envs/dev_compass_1.2.0/lib/python3.10/site-packages (from compass==1.2.0a9) (3.6.17)
Requirement already satisfied: h5py in /global/cfs/cdirs/fanssie/users/hoffman2/mambaforge/envs/dev_compass_1.2.0/lib/python3.10/site-packages (from compass==1.2.0a9) (3.10.0)
Requirement already satisfied: ipython in /global/cfs/cdirs/fanssie/users/hoffman2/mambaforge/envs/dev_compass_1.2.0/lib/python3.10/site-packages (from compass==1.2.0a9) (8.20.0)
INFO: pip is looking at multiple versions of compass to determine which version is compatible with other requirements. This could take a while.
ERROR: Could not find a version that satisfies the requirement jigsawpy (from compass) (from versions: none)
ERROR: No matching distribution found for jigsawpy
```

**Comment by xylar** [comment]
@matthewhoffman, thank you for testing.

A more efficient process for debugging is to either look at the `conda/logs/bootstrap.log` mentioned in the output or to use the `--verbose` flag so all output goes to the terminal.

That being said, I can't tell from the output you posted what is wrong.  The only clue is:
```
ERROR: Could not find a version that satisfies the requirement jigsawpy (from compass) (from versions: none)
```
which I have not seen before.

**Comment by xylar** [comment]
I'm able to reproduce the problem on Perlmutter by reproducing your workflow.  I think I understand -- I think the environment just needs to be recreated because there's a conflicting jigsawpy in there already.  Please try again with `--recreate` (which is always the first troubleshooting step when `./conda/configure_compass_env.py` fails).

One thing I noticed is that it doesn't look like you actually cd into `../jigsaw_submodule` but maybe that's an omission.  The other is that `git checkout` will leave you on a detached head so I would do a hard reset instead:
```
git worktree add ../jigsaw_submodule
git fetch --all -p
cd ../jigsaw_submodule
git reset --hard xylar/add-jigsaw-as-submodule
```

**Comment by xylar** [comment]
@matthewhoffman, I think your issue is likely different from @cbegeman's in that I don't expect you would have an empty `jigsaw-python` submodule (let me know if that's wrong, however!).  I think the issue is that you can't use the same conda environment before and after the switch to the jigsaw-python submodule (though I'm not 100% clear on why) and you instead need to `--recreate`.  I have tried to facilitate this by adding the `alpha` version back into the name of the conda environment.  I don't think I remove that intentionally in the first place.  Since the alpha version gets bumped here, a new conda environment should be created unless you specify the name of an existing environment with `--env_name`.

**Review by matthewhoffman — APPROVED** [review]
@xylar , thank you for explaining the issue I ran into.  I figured it was something straightforward like that, but I hadn't thought to try the `--recreate` flag (or those other informative flags for debugging).  I tried again with the `--recreate` flag and was indeed able to successfully create a conda env.  I then confirmed I could generate a mesh with a landice test case that uses jigsaw.  

I was going to suggest adding something to the docs about the possible need for the `--recreate` flag that first time a developer uses this version of the code, but I see that my situation my have been a fluke given your intentional bumping of the env version, and it's probably not helpful in the long run to have extra info in the docs just for this one-time transition.  We may need to socialize the need to init the new submodule, but I think that's not a big hassle. 

Thanks for taking care of the workflow changes needed to accommodate this.

**Comment by xylar** [comment]
@matthewhoffman, I agree with you that I expect this to be a one-time problem but one that we will need to make developers aware of as it comes up.  Thanks for trying this out again.  I hope that changing default environment name avoids some of the hassle you saw but I imagine others will also see weirdness on occasion because the will be using `--env_name` with an environment that already exists.

**Comment by xylar** [comment]
@cbegeman, thanks so much for your review and testing!

**Pull request merged**